kubernetes/src/main/java/com/linecorp/armeria/client/kubernetes/endpoints/KubernetesEndpointGroup.java

```diff
@@ -224,21 +251,47 @@ public static KubernetesEndpointGroupBuilder builder(Config kubeConfig) {
         this.portName = portName;
         this.nodeAddressFilter = nodeAddressFilter;
         this.autoClose = autoClose;
-        nodeWatch = watchNodes();
-        serviceWatch = watchService();
+        watchJob(this::watchNode);
+        watchJob(this::watchService);
+    }
+
+    private void watchService() {
+        final Watch oldServiceWatch = serviceWatch;
+        if (oldServiceWatch != null) {
+            oldServiceWatch.close();
+        }
+
+        if (closed) {
+            return;
+        }
+        final Watch newServiceWatch;
+        try {
+            newServiceWatch = doWatchService();
+        } catch (Exception e) {
+            logger.warn("[{}/{}] Failed to start the service watcher.", namespace, serviceName, e);
+            return;
+        }
+        // Recheck the closed flag because the doWatchService() method may take a while.
+        if (closed) {
+            newServiceWatch.close();
+        } else {
+            serviceWatch = newServiceWatch;
+            logger.info("[{}/{}] Service watcher is started.", namespace, serviceName);
+        }
     }
 
     /**
      * Watches the service. {@link Watcher} will retry automatically on failures by {@link KubernetesClient}.
      */
-    private Watch watchService() {
+    private Watch doWatchService() {
         final Watcher<Service> watcher = new Watcher<Service>() {
             @Override
             public void eventReceived(Action action, Service service0) {
                 if (closed) {
                     return;
                 }
 
+                numServiceFailures = 0;
```

Comment:
This is always reset when a watcher fails. So numServiceFailures will always be 0 or 1.

Reply:
Isn't this reset when a watcher succeeds? I'm afraid I may not understand your intention.

Reply:
Ah, that's correct. Please forget about the comment. 😓